jax/image/patches.py

```diff
@@ -0,0 +1,118 @@
+  replaced with image patches, so instead of a single `"C"` dimension, there is,
+  for example, three dimensions `"chw". The exact order of these dimensions is
+  `"c" + ''.join(c for c in rhs_spec if c not in 'OI')`, where
+  `rhs_spec == dimension_numbers[1]`.
+
+  Docstring below adapted from `jax.lax.conv_general_dilated`.
+
+  See Also:
+    https://www.tensorflow.org/xla/operation_semantics#conv_convolution
+
+  Args:
+    lhs: a rank `n+2` dimensional input array.
+    filter_shape: a sequence of `n` integers, representing the receptive window
+      spatial shape in the order as specified in
+      `rhs_spec = dimension_numbers[1]`.
+    window_strides: a sequence of `n` integers, representing the inter-window
+      strides.
+    padding: either the string `'SAME'`, the string `'VALID'`, or a sequence of
+      `n` `(low, high)` integer pairs that give the padding to apply before and
+      after each spatial dimension.
+    lhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `lhs`. LHS dilation
+      is also known as transposed convolution.
+    rhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `rhs`. RHS dilation
+      is also known as atrous convolution.
+    dimension_numbers: either `None`, or a 3-tuple
+      `(lhs_spec, rhs_spec, out_spec)`, where each element is a string
+      of length `n+2`. `None` defaults to `("NCHWD..., OIHWD..., NCHWD...")`.
+    precision: Optional. Either ``None``, which means the default precision for
+      the backend, or a ``lax.Precision`` enum value (``Precision.DEFAULT``,
+      ``Precision.HIGH`` or ``Precision.HIGHEST``).
+
+  Returns:
+    A `2n+2` array containing the image patches where output channels (`"C"`)
+    would usually be. For example if
+    `dimension_numbers = ("NcHW", "OIwh", "CNHW")`, the output has dimension
+    numbers `"CNHW" = "{cwh}NHW"`.
+
+  """
+  if dimension_numbers is None:
+    dimension_numbers = _get_default_dimension_numbers(len(filter_shape))
+
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+
+  filter_shape = tuple(filter_shape)
+  spatial_size = np.prod(filter_shape)
+  n_channels = lhs.shape[lhs_spec.index('C')]
+
+  # Move separate `lhs` spatial locations into separate `rhs` channels.
+  rhs = jnp.eye(spatial_size, dtype=lhs.dtype).reshape(filter_shape * 2)
+
+  rhs = rhs.reshape((spatial_size, 1) + filter_shape)
+  rhs = np.tile(rhs, (n_channels,) + (1,) * (rhs.ndim - 1))
+  rhs = np.moveaxis(rhs, (0, 1), (rhs_spec.index('O'), rhs_spec.index('I')))
+
+  out = lax.conv_general_dilated(
+      lhs=lhs,
+      rhs=rhs,
+      window_strides=window_strides,
+      padding=padding,
+      lhs_dilation=lhs_dilation,
+      rhs_dilation=rhs_dilation,
+      dimension_numbers=dimension_numbers,
+      precision=precision,
+      feature_group_count=n_channels
+  )
+
+  c = out_spec.index('C')
```

Comment:
@shoyer I am returning non-flat patches here, but after thinking about it more, I think this may actually be premature (and afaik potentially costly e.g on TPUs), as for example in my usecase (implementing a locally-connected layer) I don't need unflattened patches. Apparently TF doesn't unflatten them either https://www.tensorflow.org/api_docs/python/tf/image/extract_patches, so I think this could be removed too and left to the user / downstream `np.sliding_window_view` function to do the unflattening - lmk what you think!

Reply:
Agreed!